documentation/design/rar-core-scenarios.md

```diff
@@ -0,0 +1,193 @@
+significantly reducing its size. By eliminating per-reference I/O for most references, RAR would see a significant performance boost.
+
+This is assuming we trust the SDK that it passes correct data and we trust the user that they don't delete or overwrite their SDK files. If this assumption is not
+valid, the mitigation would be to store and check the timestamp of each individual file. We would still benefit from smaller on disk caches, being able to store only
+the timestamp and not assembly name for intact SDK references, but the hot scenario wouldn't get any faster than today.
+
+### Treat NuGet references as immutable
+
+NuGet references live in the NuGet cache which is conceptually immutable. If RAR takes advantage of this, it can eliminate timestamp checks for NuGet references as
+well. The risk is higher than for SDK references because overwriting files in the NuGet cache is commonly used as a cut-the-corner workaround. The benefit is smaller
+because the number of NuGet references is typically lower. The proposal is to use the immutability assumption by default, and provide an opt-out mechanism, likely in
+the form of a new `PackageReference` metadatum for the user to indicate that they plan to update references in the package.
+
+### Don't load the per project disk cache when not needed
+
+As described above, the on disk cache is not adding any value in the hot scenario because its contents already lives in the in-memory cache. The proposal is to
+load it lazily only when (and if) RAR runs into an assembly that does not have a record in the in-memory cache. In developer inner loop, when the same solution is
+built over and over again, the cache would typically not be loaded at all, unless the developer makes a change that actually changes the dependency graph.
+
+As for saving the per-project cache, the logic would stay the same as it is today. If the in-memory cache is marked _dirty_ because it's not identical to what had
+been read from the cache, it would be written back to disk. As a special case, the cache would be considered dirty if the file does not exist at all. This is to
+support scenarios like rebuilding a project by a _hot_ RAR. RAR would be able to satisfy everything from memory, but because rebuild has cleaned the intermediate
+directory, the state would not otherwise be persisted.
+
+### Don't use the SDK disk pre-cache
+
+The idea of pre-generated on-disk cache is sound. For the `ExternallyResolved` SDK assemblies specifically, though, it effectively duplicates the information already
+present in `FrameworkList.xml`. That is, it maps assembly paths to assembly names. If the need arises we may want to re-design the pre-cache to remove the major
+drawback that it duplicates itself into all per-project caches. Cold RAR would load both caches and combine their contents (currently it's either or). Until then,
```

Comment:
Is this true when targeting .NET Framework? I think that's where we'd see a win there. Though the reference assemblies aren't in the .NET (Core) SDK itself, so maybe we could separately update the netfx reference assembly packages to have a cache (and update RAR to not push it to the projects).

Reply:
That could work. Or we could enhance `FrameworkList.xml` which, as David Kean has pointed out, ships with netfx ref assemblies already, to also include dependencies. Now it has versions and public key tokens.

I generally consider projects targeting netfx as lower priority for this work. Not because they are not widely used, but because the number of references is so much lower so the perf is usually not that bad.